File: frontend/src/lib/modules.js
/**
 * Frontend mirror of the backend module registry (backend/app/modules.py).
 *
 * The SERVER is the authority — it validates PUT /api/auth/me/modules and
 * gates routers with require_module. This mirror only drives UI: navigation
 * filtering, the onboarding questionnaire, and the Settings toggles.
 * `user.enabled_modules` arrives already resolved from GET /api/auth/me.
 */

export const MODULES = [
  // Core — always on, not shown as toggles
  { key: "dashboard", label: "Dashboard", core: true },
  { key: "accounts", label: "Accounts", core: true },
  { key: "contacts", label: "Contacts", core: true },
  { key: "expenses", label: "Expenses", core: true },
  { key: "obligations", label: "Money Flow", core: true },
  { key: "net_worth", label: "Net Worth", core: true },
  // Optional
  { key: "loans", label: "Loans & Lending", description: "Money lent or borrowed with interest, EMIs, collateral" },
  { key: "property", label: "Property Deals", description: "Land/plot deals, site plans, buyers and sellers" },
  { key: "partnerships", label: "Partnerships", description: "Shared ventures with profit/loss splitting" },
  { key: "beesi", label: "Beesi / Committee", description: "Rotating committee savings (chit funds)" },
  { key: "assets", label: "Assets", description: "Gold, silver, vehicles, FDs, stocks and other holdings" },
  { key: "forecast", label: "Forecast & Liquidity", description: "Cash-flow forecasting and recurring transactions" },
  { key: "expense_analytics", label: "Expense Analytics", description: "Charts and trends over your spending" },
  { key: "reconciliation", label: "Reconciliation", description: "Cross-check ledgers against module records" },
  { key: "reports", label: "Reports", description: "PDF / Excel exports of your data" },
  { key: "chatbot", label: "AI Assistant", description: "Ask questions about your finances in plain language" },
];

export const OPTIONAL_MODULES = MODULES.filter((m) => !m.core);
export const CORE_KEYS = MODULES.filter((m) => m.core).map((m) => m.key);

/** Does this user have the module? null/undefined enabled_modules = all. */
export function hasModule(user, key) {
  if (!key) return true; // untagged nav items are always visible
  const enabled = user?.enabled_modules;
  if (enabled == null) return true;
  return enabled.includes(key);
}

/**
 * Onboarding questionnaire: plain-language questions -> module keys.
 * Answering "yes" adds the modules; the default set is always included.
 */
export const QUESTIONNAIRE = [
  {
    id: "lending",
    question: "Do you lend money to (or borrow from) people — with interest, EMIs or collateral?",
    modules: ["loans"],
  },
  {
    id: "property",
    question: "Do you deal in property — land, plots, flats — buying or selling?",
    modules: ["property"],
  },
  {
    id: "partners",
    question: "Do you run ventures or deals jointly with partners, splitting profits?",
    modules: ["partnerships", "property"],
  },
  {
    id: "beesi",
    question: "Are you part of a beesi / committee (rotating group savings)?",
    modules: ["beesi"],
  },
  {
    id: "planning",
    question: "Do you want cash-flow forecasting, recurring transactions and downloadable reports?",
    modules: ["forecast", "reports", "reconciliation"],
  },
];

/** Modules every new account starts with (mirror of DEFAULT_SIGNUP_MODULES). */
export const DEFAULT_MODULES = [...CORE_KEYS, "assets", "expense_analytics"];

File: backend/app/modules.py
"""Module registry — the single source of truth for feature entitlements.

Every feature area ("module") a user can have is declared here. The frontend
mirrors this list (frontend/src/lib/modules.js) for navigation/questionnaire,
but the SERVER is the authority: `users.enabled_modules` is validated against
this registry, and optional-module routers are gated with
`require_module(<key>)` (app/dependencies.py).

Semantics of `users.enabled_modules`:
  - None  -> ALL modules enabled (legacy/grandfathered accounts, incl. the
             platform admin). Explicit lists are only written by signup /
             the questionnaire / the Settings page.
  - list  -> exactly these modules (core modules are always force-included
             by validate_module_keys, so a stored list is never missing them).

Household guests inherit the tenant OWNER's modules — a guest of your books
sees your feature set, not their own.
"""

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Module:
    key: str
    label: str
    description: str
    core: bool = False               # core modules are always on for everyone
    suggested_with: tuple = field(default_factory=tuple)  # soft hint for the UI, not enforced


MODULE_REGISTRY: dict[str, Module] = {m.key: m for m in [
    # ── Core: every user gets these ─────────────────────────────────────────
    Module("dashboard", "Dashboard", "Overview of your finances", core=True),
    Module("accounts", "Accounts", "Bank/cash accounts and their ledger", core=True),
    Module("contacts", "Contacts", "People and businesses you deal with", core=True),
    Module("expenses", "Expenses", "Daily expense tracking with categories and limits", core=True),
    Module("obligations", "Money Flow", "Money you owe or are owed (udhaar)", core=True),
    Module("net_worth", "Net Worth", "Combined view of what you own and owe", core=True),
    # ── Optional: chosen at signup / in Settings ────────────────────────────
    Module("loans", "Loans & Lending", "Money lent or borrowed with interest, EMIs, collateral"),
    Module("property", "Property Deals", "Land/plot deals, site plans, buyers and sellers"),
    Module("partnerships", "Partnerships", "Shared ventures with profit/loss splitting",
           suggested_with=("property",)),
    Module("beesi", "Beesi / Committee", "Rotating committee savings (chit funds)"),
    Module("assets", "Assets", "Gold, silver, vehicles, FDs, stocks and other holdings"),
    Module("forecast", "Forecast & Liquidity", "Cash-flow forecasting and recurring transactions"),
    Module("expense_analytics", "Expense Analytics", "Charts and trends over your spending"),
    Module("reconciliation", "Reconciliation", "Cross-check ledgers against module records"),
    Module("reports", "Reports", "PDF / Excel exports of your data"),
    Module("chatbot", "AI Assistant", "Ask questions about your finances in plain language"),
]}

CORE_MODULE_KEYS = frozenset(k for k, m in MODULE_REGISTRY.items() if m.core)
ALL_MODULE_KEYS = frozenset(MODULE_REGISTRY)

# What a fresh signup gets before/without the questionnaire.
DEFAULT_SIGNUP_MODULES = sorted(CORE_MODULE_KEYS | {"assets", "expense_analytics"})


def validate_module_keys(keys: list[str]) -> list[str]:
    """Validate a client-supplied module list against the registry.

    Unknown keys raise ValueError; core modules are force-included so a user
    can never lock themselves out of the basics. Returns a sorted, de-duped
    list ready to store in users.enabled_modules.
    """
    unknown = set(keys) - ALL_MODULE_KEYS
    if unknown:
        raise ValueError(f"Unknown module keys: {sorted(unknown)}")
    return sorted(set(keys) | CORE_MODULE_KEYS)


def effective_modules(owner_enabled: list[str] | None) -> list[str]:
    """Resolve a tenant owner's stored value to the actual enabled set."""
    if owner_enabled is None:
        return sorted(ALL_MODULE_KEYS)
    return sorted(set(owner_enabled) | CORE_MODULE_KEYS)
